assert/assertions.go

```diff
@@ -2052,23 +2060,83 @@ func EventuallyWithT(t TestingT, condition func(collect *CollectT), waitFor time
 	}
 }
 
+// EventuallySync asserts that a given condition will be met before waitFor time
+// plus the time taken for the last call to condition to return. The condition
+// is considered met when the condition function does not report failures on the
+// CollectT passed to it. The supplied CollectT collects all errors from one
+// tick (if there are any) and if the condition is not met before EventuallySync
+// returns, then the collected errors of the last tick are copied to t. The
+// condition function is called synchronously immediately and then every tick
+// duration after that. EventuallySync will adjust the time interval or drop
+// ticks to make up for slow condition functions.
+//
+//	externalValue := false
+//	go func() {
+//		time.Sleep(8*time.Second)
+//		externalValue = true
+//	}()
+//	assert.EventuallySync(t, func(c *assert.CollectT) {
+//		// add assertions as needed; any assertion failure will fail the current tick
+//		assert.True(c, externalValue, "expected 'externalValue' to be true")
+//	}, 10*time.Second, 1*time.Second, "external state has not changed to 'true'; still false")
+func EventuallySync(t TestingT, condition func(*CollectT), waitFor, tick time.Duration, msgAndArgs ...interface{}) bool {
+	deadline := time.Now().Add(waitFor)
+
+	tickerCh := make(chan time.Time)
+	close(tickerCh)
+	ticker := (<-chan time.Time)(tickerCh)
+	if tick > 0 {
+		timeTicker := time.NewTicker(tick)
+		defer timeTicker.Stop()
+		ticker = timeTicker.C
+	}
+
+	for {
+		collect := new(CollectT)
+
+		wait := make(chan struct{})
+		go func() {
+			defer close(wait)
+			condition(collect)
+		}()
+		<-wait
+
+		if !collect.failed() {
+			return true
+		}
+
+		<-ticker
+
+		if time.Now().After(deadline) {
```

Comment:
I wonder if we should still support failing the test immediately when `deadline` elapses, even if we're in the middle of calling the condition. We'd effectively just keep the goroutine above that I commented on and `select` over the done channel + the timeout channel. Maybe this is what was meant by leaking a goroutine though.

If we don't do that, I question the value of the API accepting a duration as a timeout because the duration isn't super closely tied with the maximum time this function might take. Maybe it would be better to just accept a maximum number of tries instead.

Reply:
> I wonder if we should still support failing the test immediately when `deadline` elapses, even if we're in the middle of calling the condition. We'd effectively just keep the goroutine above that I commented on and `select` over the done channel + the timeout channel. Maybe this is what was meant by leaking a goroutine though.

This makes the callback asynchronous (sometimes) and puts us back in the situation of having to allow it to potentially leak. It was always asking for trouble. I tried a similar fix in #1653 but I don't like it.

> If we don't do that, I question the value of the API accepting a duration as a timeout because the duration isn't super closely tied with the maximum time this function might take. Maybe it would be better to just accept a maximum number of tries instead.

This is an excellent point. I hadn't considered that synchronous behaviour would suit a different function signature better. `times int` is far more natural than `waitFor time.Duration`. I've updated the PR.